miner/remote_proposals.go

```diff
@@ -175,39 +178,44 @@ func (pb *RemoteProposalBuilder) build(
 
 	for _, signer := range signers {
 		nodeId := signer.signer.NodeID()
-		proposal, nonce, err := pb.proposalSvc.Proposal(ctx, layer, nodeId)
-		if err != nil {
-			pb.logger.Error("get partial proposal", zap.Error(err))
-			continue
-		}
-		if proposal == nil {
-			// this node signer isn't eligible this epoch, continue
-			pb.logger.Info("node not eligible on this layer. will try later")
-			continue
-		}
 
 		var proofs map[types.LayerID][]types.VotingEligibility
-		if proposal.Ballot.EpochData != nil {
+		if layer.FirstInEpoch() {
+			slots, nonce, err := pb.proposalSvc.CalculateEligibilitySlotsFor(ctx, nodeId, epoch)
+			if err != nil {
+				pb.logger.Error("calculate eligibility slots error", zap.Error(err))
+				continue
+			}
```

Comment:
What if the node is down during the first layer? It's safer to calculate slots whenever the `eligibilities[nodeId]` is vacant. Perhaps you could fetch slots and calculate eligibilties here: https://github.com/spacemeshos/go-spacemesh/blob/913a87a9e7cef705dc8fae638cd79c3f690f8f71/miner/remote_proposals.go#L139-L143